### ISSUE TITLE
Raspberry Pi Support

### ISSUE THREAD

**Issue by tresf** [qzind/tray#50]
We've had an increasing amount of requests to support a [headless, print-server environment](https://github.com/qzind/tray/issues/51) and one repeating question is whether or not we support the Raspberry Pi platform for printing.

![screen shot 2016-05-07 at 1 25 48 pm](https://cloud.githubusercontent.com/assets/6345473/15093468/3e626484-1457-11e6-8aa9-22008c665477.png)

Although Raspberry Pi is an ARM-based device, [Oracle provides official support for Java running on it using Ubuntu 14.04 and higher](http://www.oracle.com/technetwork/articles/java/raspberrypi-1704896.html).  So if Oracle supports it, we can too. (albeit NOT headless, for now)

We've two Raspberry Pi 3's (Model B), 1GB Ram, 32GB Storage on their way.  We'll document/discuss the success and failures here.  For those already testing this, please post your progress here.

Note, QZ Tray 2.0.0 requires [`512MB` free memory](https://github.com/qzind/tray/issues/27) for HTML and PDF printing.

**Comment by robertcasto** [comment]
Do you need us to do some testing?

I have a Model B 1GB Raspebbery Pi 2. It would take us a bit to get it
setup, but we could look into it. What printing device would you want it to
be able to communicate with?

Robert

On Sat, May 7, 2016 at 1:19 PM, Tres Finocchiaro <notifications@github.com>
wrote:

> We've had an increasing amount of requests to support a headless,
> print-server environment and one repeating question is whether or not we
> support the Raspberry Pi platform for printing.
>
> Although Raspberry Pi is an ARM-based device, Oracle provides official
> support for Java running on it using Ubuntu 14.04 and higher. So if Oracle
> supports it, we can too.
>
> We've two Raspberry Pi 3's (Model B), 1GB Ram, 32GB Storage on their way.
> We'll document/discuss the success and failures here. For those already
> testing this, please post your progress here.
>
> Note, QZ Tray 2.0.0 requires 512MB free memory
> <https://github.com/qzind/tray/issues/27> for HTML and PDF printing.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/qzind/tray/issues/50>
>



-- 

Robert Casto
p: 1-513-847-4942
c: 1-513-886-1550
e: casto.robert@gmail.com

**Comment by tresf** [comment]
@robertcasto no need at this point.  @lite1979 and @klabarge are quite advanced with the Linux platform.  We'll ping you if need be.

**Comment by klabarge** [comment]
**Hardware:**  Raspberry Pi 3 (Model B), 1GB Ram, 32GB
**OS:** Raspbian Jessie

The tray icon is not ideal

  ![image](https://cloud.githubusercontent.com/assets/12505463/15236643/bd84533a-1894-11e6-8d56-70831eca3b2b.png)


Installing QZ Tray on the Raspberry Pi is very similar to Ubuntu:

#### Basic Steps
1. Install Java: `sudo apt-get install oracle-java8-jdk`
1. Download QZ Tray
1. Install QZ Tray: ` sudo bash qz-tray-2.0.0.run`

#### Additional Steps
1. Install Iceweasel (Firefox): `sudo apt-get install iceweasel` 
1. Install JavaFX or OpenJFX

#### JavaFX

JavaFX is required for HTML printing.  JavaFX is normally bundled with Java 8, but as of JDK 8u33 [JavaFX support has been removed](http://www.oracle.com/technetwork/java/javase/jdk-8u33-arm-relnotes-2406696.html#CACHGFJC) on the ARM bundle.

It looks like we can install [OpenJFX](https://wiki.openjdk.java.net/display/OpenJFX/OpenJFX+on+the+Raspberry+Pi) to get this working. I will take a swing at this and provide installation steps.

**Comment by dsanders11** [comment]
@klabarge, using those steps, does QZ Tray work perfectly on the RPi3? Can JavaFX or OpenJFX be skipped if not printing HTML?

**Comment by klabarge** [comment]
> using those steps, does QZ Tray work perfectly on the RPi3

@dsanders11, we haven't done a ton of bench marking with the RPi3, but our sample page/prints work well. We know of one client live with this setup. On Raspbian Jessie, the tray icon isn't the correct size.

> Can JavaFX or OpenJFX be skipped if not printing HTML?

Correct, JavaFX is only needed for HTML printing.  All other printing techniques work.

What type of printing do you plan on using?

**Comment by dsanders11** [comment]
> What type of printing do you plan on using?

Just some ZPL printing, possibly EPL in the future. We don't really need to use the RPi3, but we have a spare one and using it would certainly make the Zebra printer more mobile than it currently is tied to a desktop computer, so that is appealing.

**Comment by klabarge** [comment]
Any raw printing should work great with this setup. The only reason this bug is still open is because:

1. The tray icon
1. JavaFX or OpenJDK
 * I need to test, provide steps (or rule it out)

If you need any testing done let me know.

**Comment by tresf** [comment]
It appears a 3rd party offers the Raspberry Pi JavaFX download.

We can use this for testing until Oracle provides official support again.

http://gluonhq.com/download/javafx-embedded-sdk/

**Comment by tresf** [comment]
> 1. The tray icon

Headless support has been added via #50, so this issue is fixed.

> 2. JavaFX 

Since this is an Oracle problem, we cannot officially support HTML on Raspberry Pi.  If someone has a chance to try the ARM SDK from [gulonhq](http://gluonhq.com/download/javafx-embedded-sdk/), please post your results.  Closing as resolved.

**Comment by dsanders11** [comment]
@tresf, FYI, it might be nice to add Raspberry Pi instructions to the wiki for 'Configure Raw Printer'. I found that it's pretty much the same as the OS X instructions (CUPS) and not like the Ubuntu Linux instructions when using Raspbian.

**Comment by welrachid** [comment]
Hey guys. Will you be able to help us With our subscription to get it running on a pie? We Are considering going rPi+iPad for Mew customers.

**Comment by tresf** [comment]
@welrachid if you plan to run headless, you'd need to be on a 2.1 beta release.  2.0 still requires the system tray to be active.  2.1 hasn't been fully regression tested against the 2.0 API, so you're likely to hit compatibility issues, but the configuration is fully supported.

**Comment by welrachid** [comment]
Actually we still run 1.9 for customers because we are waiting for html support on the 2.1
I Cannot find Any roadmap on the website

PS. I know 1.9 is unsupported. We just had issues with 2.0 on customers printers where it printes rubbish AS i recall

**Comment by tresf** [comment]
QZ Tray 1.9 branch was never tested on Raspberry Pi, but if it works, you will need to run a graphical desktop environment as well.

If you find compatibility problems, you are welcome to issues a Pull Request and we can assist in the building process.

In regards to the 2.1 roadmap, we've hit some very long R&D milestones as we face limitations with the Java framework. Most notably the Java System Tray and the Live Printer Status.  Those two features have delayed the release of 2.1.   We should have 2.1 in stable form later this summer.  Sorry for the delay.

**Comment by welrachid** [comment]
Hi tres
No no.. we are currently not running with a raspberry pi. i was just wondering that if wanted to make the move to it, how would it be. we definitely will wait for 2.1 before we make that move but until then it would be nice to make some tests before that.
Is it correctly understood that the ipad running our server would then have to make a local request via example:
wss://192.168.x.x:port/ to be able to print?
while our auto-printing feature would be running directly on the raspberry pi in some sort of cronbased job retrieving or some other simple software we make ourselves?

**Comment by tresf** [comment]
> Is it correctly understood that the ipad running our server would then have to make a local request via example: wss://192.168.x.x:port/ to be able to print?

Correct, but due to the certificate restrictions on iPads, QZ Tray must be shimmed with a trusted SSL cert.  We generally recommend leveraging LetsEncrypt.org for that, if the organization is OK with rapidly expiring certs or if a wildcard cert is available, it can be converted to be shimmed at QZ Tray startup.  I would recommend getting this all working on a supported Desktop version first.

https://qz.io/wiki/print-server#remove-localhost-bindings

The relevant part is here:

> Optional: Organizations wishing to use a Trusted CA SSL certificate can do so by [manually shimming it](https://github.com/qzind/tray/pull/89#issue-169268163) into the software.

For clients running 1.9, we would recommend they first use 2.0 for key/cert conversion and then copy the necessary files to the print-server.  Note also, there's a bug with 2.0.3 and print servers, so [2.0.3-3 beta](https://github.com/tresf/tray/releases/tag/v2.0.3-3) or higher is needed.

**Comment by welrachid** [comment]
OK. Most important for us is that we dont have to update the qz tray each 3 months.
We already use letsencrypt for our hosting service..
i'll get back to you later this summer. thx

**Comment by tresf** [comment]
> Most important for us is that we dont have to update the qz tray each 3 months.

`certbot` and a post-install task should do the trick, per https://github.com/qzind/tray/issues/73#issuecomment-237165577.

```bash
[certbot ...] --posthook "sudo qz-tray-2.x.x.run -- \"-y '' path/to/fullchain.pem path/to/privkey.pem\""
```

**Comment by dsanders11** [comment]
@tresf, @klabarge, I updated the Raspberry Pi I have QZ Tray on last night to Raspberry PIXEL (see instructions [here](https://www.raspberrypi.org/blog/introducing-pixel/)) and have found that QZ Tray is *significantly* slower. Is this anything you've seen or run into, or could you update your Raspberry Pi test machine to see if it reproduces?

Running QZ Tray from a terminal window takes ~20-30 seconds before there's even any output. The CPU monitor in the tray isn't showing any kind of CPU spike or anything that may explain the extreme slowness. I can try to debug more later if there are any suggestions. Logs aren't that helpful since there's that long delay before it even starts up and puts out any logging.

**Comment by tresf** [comment]
Hmm... we haven't significantly changed our startup logic, just a Jetty websocket listener.  We do some hackish stuff with the UI in 2.0 for Ubuntu tray support, but I can't think of anything off-hand.  We'll have to do some benchmarking and get back.

**Comment by dsanders11** [comment]
@tresf, I didn't change versions of QZ Tray, just updated the Raspbian system. So still running v2.0.4, so I don't think it's anything you've changed, just that something is causing it to be slower.

It's also slower for opening connections, taking ~2-3 seconds where as it used to be nearly instant.

**Comment by tresf** [comment]
@dsanders11 thanks for clarification.  What JVM are you running?

**Comment by dsanders11** [comment]
```
java version "1.7.0_131"
OpenJDK Runtime Environment (IcedTea 2.6.9) (7u131-2.6.9-2~deb8u1+rpi1)
OpenJDK Zero VM (build 24.131-b00, interpreted mode)
```

**Comment by tresf** [comment]
Thanks.  Do you have objections to testing the Oracle flavor, whatever's the latest? (e.g. 1.8.0_x)  We have a support contract with them so if this turns out to be an upstream bug (anything but JavaFX, they pulled support) we can open a service request with them.

**Comment by tresf** [comment]
@dsanders11 just ran into this StackOverflow article that explains a bit about the performance differences on Pi.  The article is dated and we haven't reproduced your findings yet, but wanted to share.

https://stackoverflow.com/a/31156106/3196753

**Comment by dsanders11** [comment]
@tresf, I'll take a look tomorrow. It is possible the JVM changed during the update, the update process updates a lot of files.

**Comment by klabarge** [comment]
@dsanders11 I've confirmed the slowness...

TL;DR OpenJDK is slow, use oracle Java

```bash
# Install Oracle Java
sudo apt-get install oracle-java8-jdk
# Force Raspbian to use Oracle Java
sudo update-alternatives --config java
# Verify
java -version
```

#### Using OpenJDK 1.8.40

1. It takes almost 3 minutes for QZ Tray to launch. Using `top` you can see a 100% CPU usage on Java when QZ Tray is launched
1. Once QZ Tray is started, the CPU usage plummets
1. Connecting a client to the webpage triggers another 100% CPU spike for 20-30 seconds until the connection is established
1. After 1 connection is successfully made, the slowness and CPU spike is no longer reproducible until QZ Tray is restarted.

#### Using Oracle Java 1.8u91

1. It takes QZ Tray less than 10 seconds to fully start
1. Clients can connect to the webpage >1 second
